### DISCUSSION TITLE
Custom `GroupStorageProvider` in Keycloak is recognized but not registered

### DISCUSSION THREAD

**Discussion by shayanarm** [keycloak/keycloak#20222] [Q&A]
I have created a custom `GroupStorageProviderFactory` and an accompanying `GroupStorageProvider` for Keycloak's Docker image (`quay.io/keycloak/keycloak:18.0.0`) version `18.0.0` as below:

    public class XylemGroupStorageProviderFactory
      implements GroupStorageProviderFactory<XylemGroupStorageProvider> {
        static String ID = "xylem-group-storage-provider";
        protected static final List<ProviderConfigProperty> CONFIG_PROPERTIES;
        static {
            CONFIG_PROPERTIES = ProviderConfigurationBuilder.create()
                    .property().name("enabled")
                    .type(ProviderConfigProperty.BOOLEAN_TYPE)
                    .defaultValue("true")
                    .add()
                    .build();
        }
    
        @Override
        public String getId() {
            return ID;
        }
    
        @Override
        public XylemGroupStorageProvider create(KeycloakSession session, ComponentModel model) {
            GroupStorageProviderModel groupModel = new GroupStorageProviderModel(model);
            groupModel.setEnabled(true);
            return new XylemGroupStorageProvider(session, groupModel);
        }
    
    
        @Override
        public List<ProviderConfigProperty> getConfigProperties() {
            return CONFIG_PROPERTIES;
        }
    }

I even went overboard to set `enabled` properties for it, but it should not have been necessary as I have a few other custom providers which are all enabled and work by default. This particular provider does show up on the Keycloak's `providers` page on the admin panel:

[![enter image description here][1]][1]

However, the factory's `create` method never gets invoked. And my custom group storage provider never comes into play. How can I enable this custom provider?

**Update:**

I also updated my Dockerfile's `entrypoint` to enable and use the extension. But I'm not sure I am feeding the correct information to the `kc.sh build` command.

    FROM quay.io/keycloak/keycloak:18.0.2 as builder
    
    ENV KC_HEALTH_ENABLED=true
    ENV KC_METRICS_ENABLED=true
    ENV KC_FEATURES=authorization,recovery-codes,declarative-user-profile
    
    COPY ./connective-keycloak-password-entropy-18.0.0.jar /opt/keycloak/providers/connective-keycloak-password-entropy-18.0.0.jar
    COPY ./argon2-password-hash-provider-2.0.1.jar /opt/keycloak/providers/argon2-password-hash-provider-2.0.1.jar
    
    FROM quay.io/keycloak/keycloak:18.0.2
    COPY --from=builder /opt/keycloak/ /opt/keycloak/
    WORKDIR /opt/keycloak
    ENTRYPOINT ["/bin/sh", "-c", "/opt/keycloak/bin/kc.sh build --db postgres --features-disabled=impersonation --health-enabled=true --spi-group-storage-xylem-group-storage-provider-enabled=true --spi-group-storage=xylem-group-storage-provider && /opt/keycloak/bin/kc.sh start -Dkeycloak.profile.feature.declarative_user_profile=enabled -Dkeycloak.profile.feature.logging=debug --proxy edge"]

  [1]: https://i.stack.imgur.com/gMpkG.png

**Comment by maxandriani** [comment]
@shayanarm You should create an component model record to your provider on postInit hook. You need to create one for each realm you want to Enable it.
